sensor/common/internalmessage/internal_message_pubsub.go

```diff
@@ -0,0 +1,80 @@
+package internalmessage
+
+import (
+	"context"
+	"sync"
+
+	"github.com/pkg/errors"
+)
+
+const (
+	SensorMessageSoftRestart = iota
+)
+
+// SensorInternalMessageCallback is the callback used by subscribers.
+type SensorInternalMessageCallback func(message *SensorInternalMessage)
+
+// SensorInternalMessage is the message data structure used by publishers and subscribers to exchange messages.
+type SensorInternalMessage struct {
+	Kind     int
+	Text     string
+	Validity context.Context
+}
+
+// IsExpired is a helper function that checks if the context already expired without blocking.
+// If the context isn't set this function will always return false.
+func (im *SensorInternalMessage) IsExpired() bool {
+
+	if im.Validity == nil {
+		return false
+	}
+
+	select {
+	case <-im.Validity.Done():
+		return true
+	default:
+		return false
+	}
+}
+
+// NewMessageSubscriber creates a MessageSubscriber.
+func NewMessageSubscriber() *MessageSubscriber {
+	return &MessageSubscriber{
+		subscribers: map[int][]SensorInternalMessageCallback{
+			SensorMessageSoftRestart: {},
+		},
+		lock: &sync.RWMutex{},
+	}
+}
+
+// MessageSubscriber is a lightweight PubSub-like component that can be used to register callbacks and publish
+// messages.
+type MessageSubscriber struct {
+	subscribers map[int][]SensorInternalMessageCallback
+	lock        *sync.RWMutex
+}
+
+// Publish a message to all subscribers.
+func (m *MessageSubscriber) Publish(msg *SensorInternalMessage) error {
```

Comment:
What do you think about passing an interface here instead of `*SensorInternalMessage`? The `MessageSubscriber` only needs the kind to manage the callbacks and we can have an interface like:

```golang
type SensorInternalMessage interface {
    IsExpired() bool
    Kind() int
}
```

Reply:
I've done this, but I'm not sure the result is any good. 

The downside of doing this, is that now the subscribers have the responsibility of not only knowing which sub-type the message has, but also do the castng and handling incorrect type errors. See code in sensor.go where the callback is registered:

```go
v, ok := (message).(*internalmessage.SensorInternalTextMessage)
if !ok {
	log.Warnf("Soft restart message has wrong type")
	return
}
```

Reply:
Yeah. Maybe it wasn't the best suggestion. I'll leave it up to you.